Kernel/KBO.cpp

```diff
@@ -346,20 +346,55 @@ struct FuncSigTraits {
   { return env.signature->getFunction(functor)->arity() == 0; } 
 
   static Signature::Symbol* getSymbol(unsigned functor) 
-  { return env.signature->getFunction(functor); } 
+  { return env.signature->getFunction(functor); }
 };
 
 
 template<class SigTraits> 
-KboWeightMap<SigTraits> KBO::weightsFromOpts(const Options& opts) const 
+KboWeightMap<SigTraits> KBO::weightsFromOpts(const Options& opts, const DArray<int>& rawPrecedence) const
 {
   auto& str = SigTraits::weightFileName(opts);
-  if (str.empty()) {
-    return KboWeightMap<SigTraits>::dflt();
-  } else if (str == SPECIAL_WEIGHT_FILENAME_RANDOM) {
-    return KboWeightMap<SigTraits>::randomized();
-  } else {
+
+  auto arityExtractor = [](unsigned i) { return SigTraits::getSymbol(i)->arity(); };
+  auto precedenceExtractor = [&](unsigned i) { return rawPrecedence[i]; };
+  auto frequencyExtractor = [](unsigned i) { return SigTraits::getSymbol(i)->usageCnt(); };
+
+  if (!str.empty()) {
     return weightsFromFile<SigTraits>(opts);
+  } else {
+    switch (opts.kboWeightGenerationScheme()) {
+    case Options::KboWeightGenerationScheme::CONST:
+      return KboWeightMap<SigTraits>::dflt();
+    case Options::KboWeightGenerationScheme::RANDOM:
+      return KboWeightMap<SigTraits>::randomized();
+    case Options::KboWeightGenerationScheme::ARITY:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(arityExtractor,
+        [](auto _, auto arity) { return arity+1; });
+    case Options::KboWeightGenerationScheme::INV_ARITY:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(arityExtractor,
+        [](auto max, auto arity) { return max-arity+1; });
+    case Options::KboWeightGenerationScheme::ARITY_SQUARED:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(arityExtractor,
+        [](auto _, auto arity) { return arity*arity+1; });
+    case Options::KboWeightGenerationScheme::INV_ARITY_SQUARED:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(arityExtractor,
+        [](auto max, auto arity) { return max*max-arity*arity+1; });
+    case Options::KboWeightGenerationScheme::PRECEDENCE:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(precedenceExtractor,
+        [](auto _, auto prec) { return prec + 1; });
+    case Options::KboWeightGenerationScheme::INV_PRECEDENCE:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(precedenceExtractor,
+        [](auto max, auto prec) { return max-prec+1; });
+    case Options::KboWeightGenerationScheme::FREQUENCY:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(frequencyExtractor,
+        [](auto _, auto freq) { return freq > 0 ? freq : 1; });
+    case Options::KboWeightGenerationScheme::INV_FREQUENCY:
+      return KboWeightMap<SigTraits>::fromSomeUnsigned(frequencyExtractor,
+        [](auto max, auto freq) { return max > 0 ? max - freq + 1 : 1; });
```

Comment:
Really like this! U see so nicely what's going on!!

Reply:
You are the teacher of templated programming here!